### ISSUE TITLE
IO transfer error timeout and system suspend-resume test fails

### ISSUE THREAD

**Issue by Vamshigopal** [thesofproject/linux#3401]
**Describe the bug**

We can see sdw IO transfer timed out when we do suspend test. followed by an alert is raised and prepare clock error.

**To Reproduce**
run suspend_stress_test -c 10

**Reproduce rate**
sporadic.
more than 2/100 cycles

Expected result
No issue on the suspend test.

Actual result
suspend test fails.

Platform: ADL-N RVP board with onboard codec ALC711 in soundwire mode.

kernel:
v5.15 chrome kernel

[ 1842.762840] soundwire_intel soundwire_intel.link.0: IO transfer timed out, cmd 3 device 1 addr 45 len 1
[ 1842.779471] Resume caused by IRQ 9, acpi
[ 1842.795331] init: tpm2-simulator main process (3858) killed by SYS signal
[ 1842.803354] PM: suspend exit
[ 1842.818521] init: tpm2-simulator main process ended, respawning
[ 1843.363613] x86/PAT: flashrom:4265 conflicting memory types fe000000-100000000 uncached-minus<->uncached
[ 1843.374211] soundwire sdw-master-0: trf on Slave 1 failed:-110 write addr 45 count 0
[ 1843.384050] x86/PAT: memtype_reserve failed [mem 0xfe000000-0xffffffff], track uncached-minus, req uncached-minus
[ 1843.393768] rt711 sdw:0:025d:0711:00: SDW_SCP_SYSTEMCTRL write failed:-110
[ 1843.395632] rt711 sdw:0:025d:0711:00: Failed to resume device: -22
[ 1843.395643] rt711 sdw:0:025d:0711:00: Slave 1 alert handling failed: -22
[ 1843.467099] tpm tpm1: A TPM error (256) occurred attempting the self test
[ 1843.468777] rt711 sdw:0:025d:0711:00: clock stop prepare failed:-110
[ 1843.479405] tpm tpm1: starting up the TPM manually
[ 1843.489348] soundwire_intel soundwire_intel.link.0: prepare clock stop failed -110
[ 1843.548401] x86/PAT: flashrom:4265 conflicting memory types fe000000-100000000 uncached-minus<->uncached
[ 1843.552310] soundwire_intel soundwire_intel.link.0: cannot enable clock stop on suspend

[dmesg_log.txt](https://github.com/thesofproject/linux/files/7952679/dmesg_log.txt)

**Comment by Vamshigopal** [comment]
cc: @bardliao @RanderWang @plbossart @kv2019i @ranj063 @sathya-nujella

**Comment by plbossart** [comment]
@Vamshigopal can you enable dynamic debug for SoundWire (and reduce the rest of the noise that's unrelated?)

options soundwire_bus dyndbg=+p
options soundwire_generic_allocation dyndbg=+p
options soundwire_cadence dyndbg=+p
options soundwire_intel_init dyndbg=+p
options soundwire_intel dyndbg=+p
options snd_soc_rt711 dyndbg=+p

Also double-check that there are no known differences between the code in drivers/soundwire in your kernel and our latest to rule out any backport issues.

Thanks!

**Comment by Vamshigopal** [comment]
Thanks @plbossart for suggestions.

Here are the logs with dynamic debug enabled.
[dmesg_dyndb1.log](https://github.com/thesofproject/linux/files/7957842/dmesg_dyndb1.log)

Checking the differences between code, so far nothing popped up. will update if any fixes are missed.

All the major fixes are taken care in backport. Since this issue is close to #2026. Below fixes are available from backport series.  
soundwire: intel: conditionally exit clock stop mode on system suspend
soundwire: intel: don't return error when clock stop failed
soundwire: bus: update Slave status in sdw_clear_slave_status

Thanks

**Comment by plbossart** [comment]
We seem to have a race condition here, after the clock stop preparation is done, the codec driver initiates a read. we'd need a trace to see when the clock is actually stopped, it looks like the read happens after the bus is stopped, which is clearly incorrect.

````
[ 4112.100014] soundwire sdw-master-0: clock stop prep/de-prep done slave:15
[ 4112.100753] rt711 sdw:0:025d:0711:00: [rt711_sdw_read] 7520 85a0 9c20 aca0 => 00000580
[ 4112.618645] soundwire_intel soundwire_intel.link.0: IO transfer timed out, cmd 3 device 1 addr 7520 len 1
[ 4112.629375] soundwire sdw-master-0: trf on Slave 1 failed:-110 write addr 7520 count 0
[ 4112.638273] Failed to set private value: 75206f <= 0583 ret=-110
[ 4113.146626] soundwire_intel soundwire_intel.link.0: IO transfer timed out, cmd 3 device 1 addr 3501 len 1
[ 4113.157360] soundwire sdw-master-0: trf on Slave 1 failed:-110 write addr 3501 count 0
[ 4113.166256] rt711 sdw:0:025d:0711:00: rt711_calibration calibration complete, ret=0
````

I think it's possible that the headset calibration deferred work is still running.

**Comment by Vamshigopal** [comment]
ack, working to collect the traces, will share by tomorrow.

**Comment by Vamshigopal** [comment]
Facing some board issues, hence the delay in collecting traces.

**Comment by Vamshigopal** [comment]
I can see after bus clock stop there is a BUS CLASH and codec driver initiates a write.

[soix_failing.txt](https://github.com/thesofproject/linux/files/7992200/soix_failing.txt)
[soix_working.txt](https://github.com/thesofproject/linux/files/7992201/soix_working.txt)

logs from: failing case-

[  555.695971] soundwire_intel soundwire_intel.link.0: vamshi bus clock stopped 0
[  555.696004] soundwire_intel soundwire_intel.link.0: Slave status change: 0x21
[  555.704184] soundwire_intel soundwire_intel.link.0: vamshi clock stopped 0
[  555.712062] rt711 sdw:0:025d:0711:00: sdw_modify_slave_status: signaling enumeration completion for Slave 1
[  555.719767] soundwire_intel soundwire_intel.link.0: vamshi can enable clock stop on suspend:0 intel_suspend_runtime
[  556.232647] soundwire_intel soundwire_intel.link.0: IO transfer timed out, cmd 2 device 1 addr 40 len 1
[  556.243174] soundwire sdw-master-0: trf on Slave 1 failed:-110 read addr 40 count 0
[  556.251760] rt711 sdw:0:025d:0711:00: SDW_SCP_INT1 (BUS_CLASH) read failed:-110
[  556.259948] rt711 sdw:0:025d:0711:00: Slave 1 initialization failed: -110
[  556.267544] vamshi inside: rt711_sdw_write

**Comment by naveen-manohar** [comment]
Is this related to https://github.com/thesofproject/linux/pull/2582, can you check commits exists in your kernel ?

**Comment by Vamshigopal** [comment]
yes, we have commits from #2582 in my kernel.

**Comment by plbossart** [comment]
@Vamshigopal I don't understand your logs, what does 'vamshi bus clock stopped 0' means? If the bus clock is stopped, the slave status cannot change. The bus clash is irrelevant, what matter is the first IO timeout.

**Comment by Vamshigopal** [comment]
@plbossart i was checking whether the clock stopped after preparing the slaves.
in this function: int sdw_bus_clk_stop(struct sdw_bus *bus)

Slave status changes only in failing cases, in passing cases slave status doesn't change after clock stops.

**Comment by plbossart** [comment]
I don't understand your answers @Vamshigopal, no idea what the cause and effect are.

Please share the diff that adds your traces, that will be more productive, and attach the full dmesg log with dynamic debug. thank you!

**Comment by Vamshigopal** [comment]
Attaching the diff that adds trace prints and full dmesg logs with dynamic debug enabled.
[tarce_diff.txt](https://github.com/thesofproject/linux/files/7997611/tarce_diff.txt)
[dmesg_full.log](https://github.com/thesofproject/linux/files/7997613/dmesg_full.log)

**Comment by plbossart** [comment]
I just don't get what test you are trying to run here:

````
[  156.553008] rfkill rfkill1: PM: rfkill_resume+0x0/0x68 returned 0 after 21 usecs
[  156.553015] sof_sdw sof_sdw: PM: calling platform_pm_resume+0x0/0x3d @ 6882, parent: 0000:00:1f.3
[  156.553045] sof_sdw sof_sdw: PM: platform_pm_resume+0x0/0x3d returned 0 after 7 usecs
[  156.553059] input input3: PM: calling input_dev_resume+0x0/0x3e @ 6882, parent: card0
[  156.553084] input input3: PM: input_dev_resume+0x0/0x3e returned 0 after 1 usecs
[  156.553088] input input4: PM: calling input_dev_resume+0x0/0x3e @ 6882, parent: card0
[  156.553111] input input4: PM: input_dev_resume+0x0/0x3e returned 0 after 1 usecs
[  156.553115] input input5: PM: calling input_dev_resume+0x0/0x3e @ 6882, parent: card0
[  156.553138] input input5: PM: input_dev_resume+0x0/0x3e returned 0 after 1 usecs
[  156.553142] input input6: PM: calling input_dev_resume+0x0/0x3e @ 6882, parent: card0
[  156.553165] input input6: PM: input_dev_resume+0x0/0x3e returned 0 after 1 usecs
[  156.553169] input input7: PM: calling input_dev_resume+0x0/0x3e @ 6882, parent: card0
[  156.553193] input input7: PM: input_dev_resume+0x0/0x3e returned 0 after 2 usecs
[  156.553217] PM: resume of devices complete after 1389.607 msecs
[  156.554450] mei_hdcp 0000:00:16.0-b638ab7e-94e2-4ea2-a552-d1c54b627f04: bound 0000:00:02.0 (ops 0xffffffffa56c27c0)
[  156.562894] PM: Finishing wakeup.
[  156.562896] OOM killer enabled.
[  156.562896] Restarting tasks ... done.
[  156.588633] soundwire_intel soundwire_intel.link.0: Slave status change: 0x40

codec signal alert (likely headset detection)

[  156.598506] Resume caused by IRQ 9, acpi
[  156.606018] rt711 sdw:0:025d:0711:00: Slave impl defined interrupt
[  156.614708] PM: suspend exit

end of resume 

[  156.623029] rt711 sdw:0:025d:0711:00: rt711_interrupt_callback control_port_stat=4
[  156.623385] soundwire_intel soundwire_intel.link.0: Slave status change: 0x20

interrupt handled, back to 'attached'

And immediately timeout on a write to SDW_SCP_SYSTEMCTRL::ClockStopPrepare

[  160.799409] soundwire_intel soundwire_intel.link.0: IO transfer timed out, cmd 3 device 1 addr 45 len 1
[  170.171908] soundwire sdw-master-0: trf on Slave 1 failed:-110 write addr 45 count 0
[  170.180626] rt711 sdw:0:025d:0711:00: SDW_SCP_SYSTEMCTRL write failed:-110
````

I suspect the timeout is generated because of an earlier write during suspend does not complete.

You have way too many traces for irrelevant devices, please try to remove some and track every write to SCP_SYSTEMCTRL.

**Comment by plbossart** [comment]
try e.g. to remove all of these logs
````
[  137.791540] usb usb3: PM: calling usb_dev_resume+0x0/0x14 @ 4380, parent: 0000:00:14.0
[  137.791542] usb usb4: PM: calling usb_dev_resume+0x0/0x14 @ 720, parent: 0000:00:14.0
````

**Comment by Vamshigopal** [comment]
> I just don't get what test you are trying to run here:

we are running suspend-resume test, This device have RT711 as headset codec, and while running the test 3.5 headset is not connected to the target device.

> You have way too many traces for irrelevant devices, please try to remove some and track every write to SCP_SYSTEMCTRL.

trace diff:
[trace_diff.txt](https://github.com/thesofproject/linux/files/8014055/trace_diff.txt)

logs:
[dmesg_2.log](https://github.com/thesofproject/linux/files/8014018/dmesg_2.log)

[  714.201895] rt711 sdw:0:025d:0711:00: SDW_SCP_SYSTEMCTRL write failed with value:9 and return:-110

not sure why it failed writing this value 9. from logs it doesn't show any write with value 9.  

> I suspect the timeout is generated because of an earlier write during suspend does not complete.

In some runs, i don't see a SDW_SCP_SYSTEMCTRL write failed:-110. error but still the timeout is generated.

**Comment by plbossart** [comment]
> [ 714.201895] rt711 sdw:0:025d:0711:00: SDW_SCP_SYSTEMCTRL write failed with value:9 and return:-110
> 
> not sure why it failed writing this value 9. from logs it doesn't show any write with value 9.

That's the value with WakeUpEnable=1 and ClockStopPrepare=1, in other words the value that shall be written for every suspend.

I would recommend you try with the topic/sof-dev branch to rule out a backport or kernel issue. I am not able to help further at this point.

**Comment by Vamshigopal** [comment]
Ack, I will give a try.

**Comment by plbossart** [comment]
@Vamshigopal Is this problem still relevant? one month since the last update already.

**Comment by Vamshigopal** [comment]
@plbossart  yes we are still hitting this issue. Back to work after one month off, so couldn't track on this.
from #3459 i can see revert of this patch https://github.com/thesofproject/linux/commit/e38f9ff63e6d403f8e52302d223e3c5c110872ee  seems to be working. will give a try.
Please suggest any experiments/fixes if you have in list.
Thanks

**Comment by plbossart** [comment]
@Vamshigopal it's likely a different issue you're having, the timeout you reported is when accessing a standard register, while the error we see is due to a workqueue accessing vendor-specific registers. Plus you reported this error *before* that ACPI patch was upstream.

**Comment by Vamshigopal** [comment]
with recent backport of our kernel(v5.15.25) , we are not able to reproduce the issue. 100 cycles passed without reproduction.
I will confirm the observation on multiple setups. 
On side note- If i update topic/sof-dev kernel , my board is not stable, i see issues with i915 driver. This might be because some of i915 patches are not upstreamed yet.

**Comment by mengdonglin** [comment]
Close this issue as 'can't reproduce'
